engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java

```diff
@@ -609,15 +573,64 @@ private ValidationContextBuilder(
 		}
 	}
 
-	private static final class BeanAndPath {
-		private final Object bean;
-		private final Path path;
-		private final int hashCode;
+	private static final class BeanGroupProcessedUnit {
 
-		private BeanAndPath(Object bean, Path path) {
+		// these fields are final but we don't mark them as final as an optimization
+		private Object bean;
+		private Class<?> group;
+		private int hashCode;
+
+		private BeanGroupProcessedUnit(Object bean, Class<?> group) {
+			this.bean = bean;
+			this.group = group;
+			this.hashCode = createHashCode();
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if ( this == o ) {
+				return true;
+			}
+			if ( o == null || getClass() != BeanGroupProcessedUnit.class ) {
+				return false;
+			}
+
+			BeanGroupProcessedUnit that = (BeanGroupProcessedUnit) o;
+
+			if ( bean != that.bean ) {  // instance equality
+				return false;
+			}
+			if ( !group.equals( that.group ) ) {
+				return false;
+			}
+
+			return true;
+		}
+
+		@Override
+		public int hashCode() {
+			return hashCode;
+		}
+
+		private int createHashCode() {
+			int result = System.identityHashCode( bean );
+			result = 31 * result + group.hashCode();
+			return result;
+		}
+	}
+
+	private static final class BeanPathMetaConstraintProcessedUnit {
+
+		// these fields are final but we don't mark them as final as an optimization
```

Comment:
Can you shed some light on this optimization?

Reply:
To be honest, it's some dark magic suggested by Sanne. It did improve one of the benchmarks significantly.

Reply:
Hum, would be nice to have some sort of explanation...

Reply:
Based on the JMM, `final` fields are safely published, which ends up incurring a memory fence.

JLS Specification of final field behavior: https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5
SO post about the memory fence: https://stackoverflow.com/questions/23527252/cost-of-using-final-fields

Reply:
Ok, it was news to me that there's a prize to pay also if this object is only used from within one thread only. Good to know :)